### ISSUE TITLE
User module error: "Group  wheel does not exist" for existing user

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29671]
From @vbotka on 2016-09-09T05:59:38Z
##### ISSUE TYPE

 Bug Report
##### COMPONENT NAME

user module
##### ANSIBLE VERSION

```
[WARNING]: log file at /var/log/ansible.log is not writeable, aborting
ansible 2.0.0.2
  config file = /home/vlado/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

inventory   = $HOME/.ansible/hosts
remote_tmp     = $HOME/.ansible/tmp
roles_path    = $HOME/.ansible/roles
pipelining = True
##### OS / ENVIRONMENT

managing:
Linux [...] 4.4.0-36-generic #55-Ubuntu SMP Thu Aug 11 18:00:59 UTC 2016 i686 i686 i686 GNU/Linux
host:
FreeBSD [...] 10.3-RELEASE FreeBSD 10.3-RELEASE #0 r297264: Fri Mar 25 02:10:02 UTC 2016     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
##### SUMMARY

Module user

```
- name: Add or update users
  user: name="{{ item.name }}"
        groups="{{ item.groups }}" append="yes"
        shell="{{ item.shell }}"
        password="{{ item.password }}"
        update_password="{{ item.update_password }}"
  with_items: freebsd_users
```

with variables

```
freebsd_users:
  - { name: "asadmin",
      groups: "asadmin, wheel",
      shell: "/bin/sh",
      password: "<sanitized>",
      update_password: "on_create" }
  - { name: "acadmin",
      groups: "acadmin, wheel",
      shell: "/usr/local/bin/bash",
      password: "<sanitized>",
      update_password: "on_create" }
```

reports error

```
TASK [vbotka.ansible-freebsd-postinstall : Add or update users] ****************
changed: [139.59.181.117] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/bin/sh', u'name': u'asadmin', u'groups': u'asadmin, wheel', u'update_password': u'on_create'})
failed: [139.59.181.117] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/usr/local/bin/bash', u'name': u'acadmin', u'groups': u'acadmin, wheel', u'update_password': u'on_create'}) => {"failed": true, "item": {"groups": "acadmin, wheel", "name": "acadmin", "password": "$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/", "shell": "/usr/local/bin/bash", "update_password": "on_create"}, "msg": "Group  wheel does not exist"}
```

Group wheel exists. 
**Error is reported for existed user (acadmin), but not for new user (asadmin) to be created.**
##### STEPS TO REPRODUCE

dtto
##### EXPECTED RESULTS

obvious
##### ACTUAL RESULTS

```
> ansible-playbook ~/.ansible/playbooks/freebsd-postinstall.yml -t freebsd_postinstall_users -C -vvv
```

```
TASK [vbotka.ansible-freebsd-postinstall : Add or update users] ****************
task path: /export/home/vlado.config/.ansible/roles/vbotka.ansible-freebsd-postinstall/tasks/config_users.yml:2
<ip-sanitized> ESTABLISH SSH CONNECTION FOR USER: freebsd
<ip-sanitized> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=freebsd -o ConnectTimeout=30 -o ControlPath=/export/home/vlado.config/.ansible/cp/ansible-ssh-%h-%p-%r ip-sanitized '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-letpzoqgfbtjtiffwnbtqcukwpgebbcq; LANG=en_US.utf-8 LC_ALL=en_US.utf-8 LC_MESSAGES=en_US.utf-8 /usr/local/bin/python2'"'"'"'"'"'"'"'"''"'"''
<ip-sanitized> ESTABLISH SSH CONNECTION FOR USER: freebsd
<ip-sanitized> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=freebsd -o ConnectTimeout=30 -o ControlPath=/export/home/vlado.config/.ansible/cp/ansible-ssh-%h-%p-%r ip-sanitized '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-egqugqsanixhpknxllfzabpcwjlcompl; LANG=en_US.utf-8 LC_ALL=en_US.utf-8 LC_MESSAGES=en_US.utf-8 /usr/local/bin/python2'"'"'"'"'"'"'"'"''"'"''
<ip-sanitized> ESTABLISH SSH CONNECTION FOR USER: freebsd
<ip-sanitized> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=freebsd -o ConnectTimeout=30 -o ControlPath=/export/home/vlado.config/.ansible/cp/ansible-ssh-%h-%p-%r ip-sanitized '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-wpdsqgtdynhyozwwqinvsaxculqqwfps; LANG=en_US.utf-8 LC_ALL=en_US.utf-8 LC_MESSAGES=en_US.utf-8 /usr/local/bin/python2'"'"'"'"'"'"'"'"''"'"''
<ip-sanitized> ESTABLISH SSH CONNECTION FOR USER: freebsd
<ip-sanitized> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=freebsd -o ConnectTimeout=30 -o ControlPath=/export/home/vlado.config/.ansible/cp/ansible-ssh-%h-%p-%r ip-sanitized '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-gnrzjowymtilatneudojwkkhakdktjyi; LANG=en_US.utf-8 LC_ALL=en_US.utf-8 LC_MESSAGES=en_US.utf-8 /usr/local/bin/python2'"'"'"'"'"'"'"'"''"'"''
<ip-sanitized> ESTABLISH SSH CONNECTION FOR USER: freebsd
<ip-sanitized> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=freebsd -o ConnectTimeout=30 -o ControlPath=/export/home/vlado.config/.ansible/cp/ansible-ssh-%h-%p-%r ip-sanitized '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-pdckzgoepvrjtrdwqypspzjbkxeuupvz; LANG=en_US.utf-8 LC_ALL=en_US.utf-8 LC_MESSAGES=en_US.utf-8 /usr/local/bin/python2'"'"'"'"'"'"'"'"''"'"''
changed: [ip-sanitized] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/bin/sh', u'name': u'asadmin', u'groups': u'asadmin, wheel', u'update_password': u'on_create'}) => {"changed": true, "invocation": {"module_args": {"_ansible_check_mode": true, "append": true, "comment": null, "createhome": true, "expires": null, "force": false, "generate_ssh_key": null, "group": null, "groups": "asadmin, wheel", "home": null, "login_class": null, "move_home": false, "name": "asadmin", "non_unique": false, "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "remove": false, "shell": "/bin/sh", "skeleton": null, "ssh_key_bits": "2048", "ssh_key_comment": "ansible-generated on srv.netng.co", "ssh_key_file": null, "ssh_key_passphrase": null, "ssh_key_type": "rsa", "state": "present", "system": false, "uid": null, "update_password": "on_create"}, "module_name": "user"}, "item": {"groups": "asadmin, wheel", "name": "asadmin", "password": "$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/", "shell": "/bin/sh", "update_password": "on_create"}}
failed: [ip-sanitized] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/usr/local/bin/bash', u'name': u'acadmin', u'groups': u'acadmin, wheel', u'update_password': u'on_create'}) => {"failed": true, "invocation": {"module_args": {"_ansible_check_mode": true, "append": true, "comment": null, "createhome": true, "expires": null, "force": false, "generate_ssh_key": null, "group": null, "groups": "acadmin, wheel", "home": null, "login_class": null, "move_home": false, "name": "acadmin", "non_unique": false, "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "remove": false, "shell": "/usr/local/bin/bash", "skeleton": null, "ssh_key_bits": "2048", "ssh_key_comment": "ansible-generated on srv.netng.co", "ssh_key_file": null, "ssh_key_passphrase": null, "ssh_key_type": "rsa", "state": "present", "system": false, "uid": null, "update_password": "on_create"}, "module_name": "user"}, "item": {"groups": "acadmin, wheel", "name": "acadmin", "password": "$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/", "shell": "/usr/local/bin/bash", "update_password": "on_create"}, "msg": "Group  wheel does not exist"}
```

Copied from original issue: ansible/ansible-modules-core#4756

**Comment by ansibot** [comment]
From @vbotka on 2016-09-09T05:59:38Z

Running the task with modified variables (wheel removed from the groups of acadmin). 

```
freebsd_users:
  - { name: "asadmin",
      groups: "asadmin, wheel",
      shell: "/bin/sh",
      password: "<sanitized>",
      update_password: "on_create" }
  - { name: "acadmin",
      groups: "acadmin",
      shell: "/usr/local/bin/bash",
      password: "<sanitized>",
      update_password: "on_create" }
```

resulted in the error **"Group asadmin does not exist"**

```
TASK [vbotka.ansible-freebsd-postinstall : Add or update users] ****************
failed: [139.59.181.117] => (item={u'password': <sanitized>, u'shell': u'/bin/sh', u'name': u'asadmin', u'groups': u'asadmin, wheel', u'update_password': u'on_create'}) => {"failed": true, "item": {"groups": "asadmin, wheel", "name": "asadmin", "password": <sanitized>, "shell": "/bin/sh", "update_password": "on_create"}, "msg": "Group asadmin does not exist"}
```

Group asadmin really doesn't exist yet. In FreeBSD the default group is user's own group which will be created together with the user.

Thus, I've removed the primary groups from the variables

```
freebsd_users:
  - { name: "asadmin",
      groups: "wheel",
      shell: "/bin/sh",
      password: "<sanitized>",
      update_password: "on_create" }
  - { name: "acadmin",
      groups: "wheel",
      shell: "/usr/local/bin/bash",
      password: "<sanitized>",
      update_password: "on_create" }
```

and the task passed.

```
TASK [vbotka.ansible-freebsd-postinstall : Add or update users] ****************
changed: [139.59.181.117] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/bin/sh', u'name': u'asadmin', u'groups': u'wheel', u'update_password': u'on_create'})
changed: [139.59.181.117] => (item={u'password': u'$6$RorpzCuGgTXtEVfk$TPDhEQzVOlDO0Op0ZraJJS6ZSvknsEPqqxrpbBDE18p.az4LHzzfs0shTBS.GMNkFv2t9BNEL4d6ECLV6aW4j/', u'shell': u'/usr/local/bin/bash', u'name': u'acadmin', u'groups': u'wheel', u'update_password': u'on_create'})
```

**Probably this whole bug report might be closed as INVALID and just update the documentation of User module: "groups parameter must not comprise user's primary group" (at least for FreeBSD).**

As a result, separate task is needed to set additional groups. Example:

```
- name: Add or update users
  user: name="{{ item.name }}"
        shell="{{ item.shell }}"
        password="{{ item.password }}"
        update_password="{{ item.update_password }}"
  with_items: freebsd_users

- name: Add users to additional groups
  user: name="{{ item.name }}"
        groups="{{ item.groups }}"
        append="yes"
  with_items: freebsd_users_groups
```

JFYI. I haven't tested if append="yes" works.

**Comment by ansibot** [comment]
From @ansibot on 2016-09-09T05:59:38Z

@grossws, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @ansibot on 2016-09-09T05:59:38Z

@ansible, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @chrnie on 2016-09-09T05:59:38Z

Try it without the space after the comma. This fixed it for me.
befor:
`groups: "asadmin, wheel",`
after:
`groups: "asadmin,wheel",`

**Comment by ansibot** [comment]
cc @sfromm
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @samdoran
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by samdoran** [comment]
This was fixed in d0e3d75381292fc7bf417598f9f6542a7fcd2d07
